### ISSUE TITLE
Geometric Multigrid for faster linear solvers

### ISSUE THREAD

**Issue by tjhei** [geodynamics/aspect#2767]
issue to track progress for GMG that @tcclevenger is working on:

- [x] simple benchmark problems
- [x] free slip
- [x] spherical geometry and free slip (with deal.II master)
- [x] relax averaging requirement
- [x] compressibility
- [ ] periodic boundary conditions
- [x] free surface
- [ ] BFBT Schur complement

**Comment by tcclevenger** [comment]
An update: 
Currently we have a matrix-free Stokes block/solver for prescribed velocity boundary (all meshes) and free-slip boundary (box only). We use GMG for velocity block. _Must use cell-wise averaging_.

Still needed:
1. Free-slip boundary for sphere/spherical shell. (current: works in 2D serial, should be simple fix for parallel, will require dealii-developer version, 3D should be logical extension)
2. Relax the requirement for cell-wise averaging. Options: store all quadrature points (not ideal) or project with higher order finite element.
3. Compressibility (should be easy for basic case, already implemented in part, but not tested)
4. Free surface
5. BFBT for schur complement
6. Melt transport

**Comment by maxrudolph** [comment]
Would it be  difficult to extend the GMG to work for periodic boundary conditions? I would be interested in using it for a 3D cartesian model with periodic x- and y- boundary conditions.

**Comment by tjhei** [comment]
> Would it be difficult to extend the GMG to work for periodic boundary conditions? I would be interested in using it for a 3D cartesian model with periodic x- and y- boundary conditions.

I was thinking about that when you posted your test case the other day. :-)
It is likely not too difficult to make it work. Conrad said he would try to do it at some time this spring. He will talk about this (and other plans) at the virtual hackathon tomorrow.

**Comment by tcclevenger** [comment]
Yes, I’ll mention this tomorrow. All the required stuff should already be in DEALII so it will hopefully not be too much work to add the functionality to aspect. I’ll keep you updated.

**Comment by tcclevenger** [comment]
@gassmoeller and @tjhei 
Here are some initial results for the viscosity projection:

-  I tested 3 different benchmark problems which have reference solutions and smooth viscosity: solkz, hollow sphere, burstedde. 

- My viscosity strategy: take quadrature defined, active mesh viscosity (given by the benchmark), project this to a `DGQ(k)` vector on the active mesh, transfer this vector to GMG level vectors, evaluate vectors at quadrature points on GMG levels. 

- For all other coefficients (I think just "density" in these benchmarks), averaging set to "none", i.e., evaluated at the quadrature points.

- The GMG stuff is unimportant for the user; it is only important to know that the viscosity now used to define the problem (on the active mesh) comes from a `DGQ(k)` representation instead of quadrature point data. 

- Results: The results are very similar for each benchmark. I'll just show the hollow_sphere since it is 3d and non-constant density. (viscosity parameter set to `m=5`)

Here we have the `u` and `p`,  `L1` and `L2` errors (and rates below) for AMG with "none" set as averaging, GMG with harmonic averaging, GMG with `DGQ(2)` and `DGQ(1)` viscosity averaging.

![hollow_sphere](https://user-images.githubusercontent.com/11428698/73298681-c1020600-41db-11ea-9cb2-9e76640dc008.png)

My takeaways:
1. AMG with no averaging has almost identical errors/rates as GMG with the `DGQ(2)` projection. 
2. GMG with `DGQ(1)` has slightly smaller errors (similar rates) to `DGQ(2)`.

This seems promising. Thoughts? What other problems would be interesting to test with?

**Comment by tjhei** [comment]
> Thoughts?

Interestingly, DGQ1 gives slightly better errors and same rates compared to DGQ2 and the reference. I thought there are cases where this is not the case and DGQ1 gives lower accuracy. @gassmoeller ?

> What other problems would be interesting to test with?

solkz (matching and non-matching) and inclusion come to mind. For Q1 you might be able to do this without worrying about over/undershoots.

**Comment by gassmoeller** [comment]
Ah, I do not remember the details of what we discussed, but I might have misled you. I checked our paper again were we made the convergence rate analysis and the numerical tests (https://academic.oup.com/gji/article-pdf/219/3/1915/30105955/ggz405.pdf). DGQ(1) for viscosity has the same rate as the bilinear least-squares fit we used in the paper, and will allow us a third order rate for velocity and second order for pressure (so an optimal rate for a Q2Q1 element). However, it will not allow us any higher rates. So if you would combine DGQ(1) with a Q3Q2 element you should still see the same (now sub-optimal) rates. So for our purposes in ASPECT DGQ(1) is absolutely enough, which should make dealing with overshoots simpler.

**Comment by tjhei** [comment]
#4251 extends supported boundary conditions.
#4217 implements Newton derivatives (first Part)

**Comment by maxrudolph** [comment]
Hi @tjhei - I tried to use this most recent version for a 3D spherical geometry problem with prescribed velocity boundary conditions and got a convergence failure during the initial refinement. The same problem does converge using the block AMG solver. Do you have any guidance on this? Also, I am using the Q2Q1 element for velocity/pressure. Should I be using a DG element for pressure with this solver? Also - is it possible to resume from a checkpoint and switch from block AMG to block GMG, or to fall back on block AMG if block GMG fails to converge?

**Comment by tjhei** [comment]
I know things are not ideal for free slip and things work for 0 velocity. I have not tested with other prescribed velocities. Do you have an easy test prm where I can see the problem?

**Comment by maxrudolph** [comment]
I have a test .prm file. But it is not easy. After the initial refinement, this model has about 500 million degrees of freedom.

The necessary viscosity profile and initial temperature profile are here:
https://github.com/maxrudolph/aspect-energy-equation/tree/master/plates-amr
The initial plate motion stage is here:
https://github.com/maxrudolph/aspect-energy-equation/tree/master/rcf-testcase/ZZ%2BS200

[convection-plates-spinup_aspect-mpiio_amr3_time0.prm.txt](https://github.com/geodynamics/aspect/files/6874170/convection-plates-spinup_aspect-mpiio_amr3_time0.prm.txt)
